torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -547,40 +543,41 @@ def _setup_flat_buffers(self) -> None:
         parameter trainability is changed (frozen or unfrozen) and
         ``_update_trainable`` is called.
         """
-
-        for device, per_rank_params in self._per_device_params.items():
-            # Only wipe the existing buckets if there are none
-            # (could be that this is called twice, when trainability changes)
-            if device not in self.buckets.keys():
-                self.buckets[device] = []
-
-            # Make parameters a view of the bucket
-            for dst_rank, params in enumerate(per_rank_params):
-                if len(params) > 0:
-
-                    # Clone the non-trainable params, if in a bucket it will get destroyed
-                    for param in filter(lambda x: not x.requires_grad, params):
+        for rank, param_groups in enumerate(self.partition_parameters()):
+            # Clone the non-trainable params, find the buffer size and dtype
+            # for the trainable params' bucket, and compile a list of the
+            # trainable params
+            buffer_size = 0
+            dtype = None
+            trainable_params = []
+            for param_group in param_groups:
+                for param in param_group["params"]:
+                    if (not _is_trainable(param)):
                         param.data = param.data.detach().clone()
```

Comment:
I also did not understand "if in a bucket it will get destroyed". Could you explain?

Reply:
@blefaudeux  are you referring to the `parameters_as_bucket_view=True` case, where `clone()` here makes the param no longer a view on the bucket?